teams/views.py

```diff
@@ -0,0 +1,116 @@
+                          create_teams_in_view, update_team_participants
+
+
+@login_required
+def change_teams(request, hackathon_id):
+    """ Page that handles the logic of automatically distributing the teams
+    for a hackathon and allows for the admin to re-arrange the team members """
+    # Redirect user if they are not admin
+    if not request.user.is_superuser:
+        return redirect(reverse('hackathon:hackathon_detail',
+                                kwargs={'pk': hackathon_id}))
+
+    edit = False
+    hackathon = Hackathon.objects.get(id=hackathon_id)
+    participants = hackathon.participants.all()
+    if len(participants) == 0:
+        teams = []
+        participants_still_to_distribute = []
+    elif len(hackathon.teams.all()) == 0:
+        teamsize = 3
+        team_sizes = sorted(choose_team_sizes(participants, teamsize))
+        if len(team_sizes) == 0:
+            return render(request, 'change_teams.html',
+                        {'num_participants': len(participants)})
+        grouped_participants, hackathon_level = group_participants(
+            participants, len(team_sizes))
+        team_levels = sorted(choose_team_levels(len(team_sizes), hackathon_level))
+        combos_without_dupes = find_all_combinations(
+            participants, team_sizes)
+        teams, leftover_participants = distribute_participants_to_teams(
+            team_sizes, team_levels, grouped_participants,
+            combos_without_dupes)
+        participants_still_to_distribute = [
+            participant for participant_groups in leftover_participants.values()
+            for participant in participant_groups]
+    else:
+        edit = True
+        teams = hackathon.teams.all()
+        team_members = [user for team in teams
+                        for user in team.participants.all()]
+        participants_still_to_distribute = [participant.to_team_member()
+                                            for participant in participants
+                                            if participant not in team_members]
+        teams = {team.display_name: [
+                    team_member.to_team_member()
+                    for team_member in team.participants.all()]
+                 for team in teams}
+
+    return render(request, 'change_teams.html', {
+        'hackathon_id': hackathon_id,
+        'num_participants': len(participants), 
+        'teams': teams,
+        'leftover_participants': participants_still_to_distribute,
+        'edit': edit,
+        })
+
+
+@login_required
+def create_teams(request):
+    if not request.user.is_superuser:
+        messages.error(request, "You do not have access to this page!")
+        return redirect(reverse('hackathon:hackathon-list'))
+    if request.method == 'POST':
+        data = request.POST
+        teams = json.loads(data.get('teams'))
+        hackathon_id = data.get('hackathon_id')
+        if data.get('_method') == 'post':
+            with transaction.atomic():
```

Comment:
Oooh! Shiny! Never seen this before.... will need to look it up to see what this does!

Reply:
It should basically roll everything back that was inserted up until it throws an exception (if it throws an exception). Here more info for reference: https://docs.djangoproject.com/en/3.1/topics/db/transactions/#controlling-transactions-explicitly

Reply:
I don't remember where I saw it, but I liked it and like now adding it if I am doing these type of insertions as we don't want teams to be partially created if there is an exception.

Reply:
Ah very cool! I saw that last night - it's something to remember for later, certainly!